sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala

```diff
@@ -34,13 +36,53 @@ import org.apache.spark.sql.rapids.GpuFileSourceScanExec
 class GpuTransitionOverrides extends Rule[SparkPlan] {
   var conf: RapidsConf = null
 
-  def optimizeGpuPlanTransitions(plan: SparkPlan): SparkPlan = plan match {
+  def optimizeGpuPlanTransitions(plan: SparkPlan): SparkPlan = {
+
+    plan match {
+      case HostColumnarToGpu(r2c: RowToColumnarExec, goal) =>
+        GpuRowToColumnarExec(optimizeGpuPlanTransitions(r2c.child), goal)
+      case ColumnarToRowExec(bb: GpuBringBackToHost) =>
+        GpuColumnarToRowExec(optimizeGpuPlanTransitions(bb.child))
+      case p =>
+        p.withNewChildren(p.children.map(optimizeGpuPlanTransitions))
+    }
+  }
+
+  def optimizeAdaptiveTransitions(plan: SparkPlan): SparkPlan = plan match {
+
     case HostColumnarToGpu(r2c: RowToColumnarExec, goal) =>
-      GpuRowToColumnarExec(optimizeGpuPlanTransitions(r2c.child), goal)
+      GpuRowToColumnarExec(optimizeAdaptiveTransitions(r2c.child), goal)
+
+    case GpuCoalesceBatches(e: GpuShuffleExchangeExec, _) =>
+      // we need to insert the coalesce batches step later, after the query stage has executed
```

Comment:
I realized this morning that we are losing some of the benefits of GpuCoalesceBatches now because we apply it to the shuffle reader rather than to the output of the shuffle, so it is running in a different place now. I think we will either need to embed the GpuCoalesceBatches logic into GpuShuffleExchangeExec or implement a new GpuCoalesceExchange operator that implements Exchange and wraps GpuShuffleExchangeExec. The reason for this is that the operator we return in newQueryStage must directly be an Exchange operator.